src/swarm/node/request/RequestStats.d

```diff
@@ -745,4 +745,320 @@ public class RequestStats
+                    hash of this instance
+
+            *******************************************************************/
+
+            public hash_t toHash ( ) /* d1to2fix_inject: const */
+            {
+                return Fnv1a.combined(this.request, this.client);
+            }
+
+            /*******************************************************************
+
+                Struct equality function. Required for use as an AA key in D2.
+
+                Params:
+                    rhs = other instance to compare against
+
+                Returns:
+                    true if rhs is equal to this instance
+
+            *******************************************************************/
+
+            mixin (genOpEquals("
+            {
+                return this.request == rhs.request && this.client == rhs.client;
+            }
+            "));
+
+            /*******************************************************************
+
+                Struct comparison function. Required for use as an AA key in D1.
+
+                TODO: can be removed, when we ditch D1.
+
+                Params:
+                    rhs = other instance to compare against
+
+                Returns:
+                    0 if rhs is equal to this instance, 1 otherwise
+
+            *******************************************************************/
+
+            mixin (genOpCmp("
+            {
+                if ( this.request == rhs.request && this.client == rhs.client )
+                    return 0;
+                else
+                    return 1;
+            }
+            "));
+        }
+
+        /// Per-client/request information map.
+        private ClientRequestInfo[ClientRequest] client_request_info;
+
+        /***********************************************************************
+
+            Called when a scheduled-for-removal request starts to be handled.
+
+            Params:
+                rq = name of request
+                client = name of client that sent the request
+                addr = remote addr/port of client
+
+        ***********************************************************************/
+
+        public void started ( cstring rq, cstring client, AddrPort addr )
+        {
+            auto now = time(null);
+            auto cl_rq = ClientRequest(client, rq);
+            auto info = cl_rq in this.client_request_info;
+            if ( info is null )
+            {
+                this.client_request_info[ClientRequest(client.dup, rq.dup)]
+                    = ClientRequestInfo.init;
+                info = cl_rq in this.client_request_info;
+            }
+            assert(info !is null);
+
+            // Warn the first time a client sends an old request version.
+            if ( !info.active_this_hour(now) )
+                logger.warn("Client '{}' on {}:{} sent an old version of request '{}'",
+                    client, addr.address_bytes, addr.port, rq);
```

Comment:
I think the comment here should be made that this is triggered only once per hour per client per request, not only the first time (as far as I can see).